src/initial_questions/inital_questions.py

```diff
@@ -28,24 +29,44 @@ def fetch_relevant_data_from_k8s_cluster(
         """Fetch the relevant data from Kubernetes cluster based on specified K8s resource in message."""
         ...
 
+    def apply_token_limit(self, text: str, token_limit: int) -> str:
+        """Reduces the amount of tokens of a string by truncating exeeding tokens.
+        Takes the template into account."""
+        ...
+
 
 class InitialQuestionsHandler:
     """Handler that generates initial questions."""
 
     chain: typing.Any
+    _model: IModel
+    _template: str
+
+    def __init__(self, model: IModel, template: str | None = None) -> None:
+        self._model = model
+        self._template = template or INITIAL_QUESTIONS_PROMPT
 
-    def __init__(
-        self,
-        model: IModel,
-    ) -> None:
-        prompt_template: str = INITIAL_QUESTIONS_PROMPT
         prompt = PromptTemplate(
-            template=prompt_template,
+            template=self._template,
             input_variables=["context"],
         )
         output_parser = QuestionOutputParser()
         self.chain = prompt | model.llm | output_parser
 
+    def apply_token_limit(self, text: str, token_limit: int) -> str:
+        """Reduces the amount of tokens of a string by truncating exeeding tokens.
+        Takes the template into account."""
+        tokenizer = tiktoken.encoding_for_model(self._model.name)
+
+        tokens_template = tokenizer.encode(self._template)
+        tokens_text = tokenizer.encode(text)
+        token_limit -= len(tokens_template)
```

Comment:
sometimes tokens_template token number can be higher than tokens_text. In this case `tokens_limit` will be negative. You need to handle that case too.

Reply:
"Sometimes tokens_template token number can be higher than tokens_text. In this case tokens_limit will be negative. You need to handle that case too."

No.

Say token_template has a len of 10, token_text has a len of 5, so as you say template>text. Let's say TokenLimit is 12.  `token_limit -= len(tokens_template)` would still not be negative. 

It would only be negative in the unlikely event of the template being longer than the limit. If the template is longer than the token limit we will not be able to send a request to the model anyhow because it does not make any sense to shorten the template.

Reply:
I have chosen an in-between way; I will now handle the case where template is longer than token limit.